Load the default config file before working with default properties

mattcaswell opened the pull request:
A config file can change the global default properties. Therefore we
must ensure that the config file is loaded before reading or amending
them.

Fixes #12565

Marked as WIP because it needs some kind of test.

levitte commented:
Why is the load config flag needed?

mattcaswell commented:
> Why is the load config flag needed?

Because without it the config file doesn't get loaded! ;-)

mattcaswell commented:
I've now updated this with a test and taken it out of WIP.

Please take a look.

t8m commented on crypto/evp/evp_fetch.c:
```diff
@@ -385,21 +386,26 @@ static int evp_set_default_properties(OPENSSL_CTX *libctx,
     return 0;
 }
 
-int EVP_set_default_properties(OPENSSL_CTX *libctx, const char *propq)
+int evp_set_default_properties_intern(OPENSSL_CTX *libctx, const char *propq,
+                                      int loadconfig)
```
It is really confusing to have evp_set_default_properties EVP_set_default_properties and evp_set_default_properties_intern. Perhaps rename evp_set_default_properties to evp_set_parsed_default_properties and (optionally) evp_set_default_properties_intern to evp_set_default_properties.

mattcaswell replied:
I renamed `evp_set_default_properties` to `evp_set_parsed_default_properties` but left `evp_set_default_properties_intern` as it is.

t8m replied:
One reason (unwritten here, my bad) why I suggested the rename of the _intern function was that the suffix is not used anywhere else - we usually use _int. But anyway it is fine to keep it as is.

mattcaswell replied:
I renamed it to use _int

mattcaswell commented:
I've rebased this and completely reworked the test so that it is a new standalone test rather than a part of evp_test. Previously it relied on the fact that evp_test used the default libctx. Meanwhile @slontis came along and changed it so that it didn't any more :-)

Please take another look.

t8m commented on test/recipes/30-test_defltfips.t:
```diff
@@ -0,0 +1,43 @@
+}
+
+$ENV{OPENSSL_CONF} = abs_path(srctop_file("test", "default.cnf"));
+ok(run(test(["defltfips_test"])), "running defltfips_test");
```
Missing EOL at EOF.

t8m commented on test/defltfips_test.c:
```diff
@@ -0,0 +1,81 @@
+    /* Must be the first test before any other libcrypto calls are made */
+    ADD_TEST(test_is_fips_enabled);
+    return 1;
+}
```
Missing EOL at EOF.

mattcaswell commented:
Two missing EOLs fixed. Please take another look.

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented:
Pushed. Thanks.
